---
title: RCCX Minigraph-Cactus pangenome summary
output: github_document
---


```{r include=FALSE}
knitr::opts_chunk$set(echo=TRUE, message=FALSE, warning=FALSE, fig.width=10)
```

```{r}
library(dplyr)
library(ggplot2)
library(knitr)
```

# GRCh38 + HPRC alleles

## Node metadata

```{r}
ninfo = read.table('rccx.grch38_hprc.mc.node_info.tsv', as.is=TRUE, header=TRUE)
```

```{r}
ninfo %>% group_by(class) %>%
  summarize(n=n(), mean.size.bp=round(mean(size), 1),
            total.bp=round(sum(size), 1)) %>%
  kable(format.args=list(big.mark=','))
```

Node classes:

- *c1*: specific to module 1
- *c2*: specific to module 1
- *cyc_l*/*cyc_r*: nodes involved in the loop back, e.g. from end of module 1 to beginning of module 2.
- *none*: uninformative nodes
- *ref*: flanking reference nodes, outside of the RCCX modules.

```{r}
ggplot(ninfo, aes(x=c1, y=c2, color=class)) +
  geom_point(alpha=.8) +
  scale_color_brewer(palette='Set1') +
  xlab('number of module 1 haplotypes') +
  ylab('number of module 2 haplotypes') +
  theme_bw()
```

```{r}
ggplot(ninfo, aes(x=c1)) +
  geom_histogram(binwidth=1) +
  xlab('number of module 1 haplotypes') +
  ylab('number of nodes') + 
  scale_x_continuous(breaks=seq(0,100,5)) + 
  theme_bw()
ggplot(ninfo, aes(x=c2)) +
  geom_histogram(binwidth=1) +
  xlab('number of module 2 haplotypes') +
  ylab('number of nodes') +
  scale_x_continuous(breaks=seq(0,100,5)) + 
  theme_bw()
```

# GRCh38 only

## Node metadata

```{r}
ninfo = read.table('rccx.grch38.mc.node_info.tsv', as.is=TRUE, header=TRUE)
```

```{r}
ninfo %>% group_by(class) %>%
  summarize(n=n(), mean.size.bp=round(mean(size), 1),
            total.bp=round(sum(size), 1)) %>%
  kable(format.args=list(big.mark=','))
```

Node classes:

- *c1*: specific to module 1
- *c2*: specific to module 1
- *cyc_l*/*cyc_r*: nodes involved in the loop back, e.g. from end of module 1 to beginning of module 2.
- *none*: uninformative nodes
- *ref*: flanking reference nodes, outside of the RCCX modules.

```{r}
ggplot(ninfo, aes(x=c1, y=c2, color=class)) +
  geom_point(alpha=.8) +
  scale_color_brewer(palette='Set1') +
  xlab('number of module 1 haplotypes') +
  ylab('number of module 2 haplotypes') +
  theme_bw()
```

```{r}
ggplot(ninfo, aes(x=c1)) +
  geom_histogram(binwidth=1) +
  xlab('number of module 1 haplotypes') +
  ylab('number of nodes') + 
  scale_x_continuous(breaks=seq(0,100,5)) + 
  theme_bw()
ggplot(ninfo, aes(x=c2)) +
  geom_histogram(binwidth=1) +
  xlab('number of module 2 haplotypes') +
  ylab('number of nodes') +
  scale_x_continuous(breaks=seq(0,100,5)) + 
  theme_bw()
```
